python/tank/bootstrap/resolver.py

```diff
@@ -255,31 +264,67 @@ def resolve_shotgun_configuration(
 
-                if self.__match_plugin_id(pc.get("plugin_ids")) or self.__match_plugin_id(pc.get("sg_plugin_ids")):
+                if self._match_plugin_id(pc.get("plugin_ids")) or self._match_plugin_id(pc.get("sg_plugin_ids")):
                     # we have a matching pipeline configuration!
 
-                    if pc["code"] == constants.PRIMARY_PIPELINE_CONFIG_NAME:
-                        if primary_config:
-                            log.warning(
-                                "More than one pipeline config detected. Will use the most "
-                                "recently updated one."
-                            )
-                        primary_config = pc
+                    if pc["project"] == self._proj_entity_dict:
+
+                        # this pipeline configuration matches our current project exactly!
+                        # alternatively, we may be in site mode, where project id is always None.
+                        # this kind of exact match takes precdence (see logic below)
+                        if pc["code"] == constants.PRIMARY_PIPELINE_CONFIG_NAME:
+                            log.debug("Primary match: %s" % pc)
+                            primary_config = pc
+                        else:
+                            user_config = pc
+                            log.debug("Per-user match: %s" % pc)
+
                     else:
-                        # user config
-                        if user_config:
-                            log.warning(
-                                "More than one user config detected. Will use the most "
-                                "recently updated one."
-                            )
-                        user_config = pc
 
-            # user pc takes precedence if available.
-            pipeline_config = user_config if user_config else primary_config
+                        # alternatively, this is a pipeline configuration record
+                        # which doesn't match directly - typically this is a
+                        # pipeline config record with project id set to None, in this
+                        # case indicating that this configuration can be used for
+                        # *any* project on the site (one config used for all projects).
+                        # this has a lower priority than the exact match above, so if
+                        # a project has specific pipeline configuration specified, this
+                        # always takes precedence.
+                        if pc["code"] == constants.PRIMARY_PIPELINE_CONFIG_NAME:
+                            primary_config_fallback = pc
+                            log.debug("Found primary fallback match: %s" % pc)
+                        else:
+                            user_config_fallback = pc
+                            log.debug("Found per-user fallback match: %s" % pc)
+
+
+            # Now select in order of priority:
```

Comment:
Technically, the code below is in reverse priority order. Any reason not to do like I suggested in previous comment and do the checks in order with `if/else`? Fine either way. Maybe I'm missing something subtle though.